pennylane_lightning/core/src/simulators/lightning_qubit/gates/GateIndices.cpp

```diff
@@ -47,4 +47,34 @@ auto generateBitPatterns(const std::vector<std::size_t> &qubitIndices,
     }
     return indices;
 }
+
+/**
+ * @brief Introduce quantum controls in indices generated by
+ * generateBitPatterns.
+ *
+ * @param indices Indices for the operation.
+ * @param num_qubits Number of qubits in register.
+ * @param controlled_wires Control wires.
+ * @param controlled_values Control values (false or true).
+ */
+void controlBitPatterns(std::vector<std::size_t> &indices,
+                        const std::size_t num_qubits,
+                        const std::vector<std::size_t> &controlled_wires,
+                        const std::vector<bool> &controlled_values) {
+    constexpr std::size_t one{1U};
+    if (controlled_wires.empty()) {
+        return;
+    }
+    std::for_each(
+        indices.begin(), indices.end(),
+        [num_qubits, &controlled_wires, &controlled_values](std::size_t &i) {
+            for (std::size_t k = 0; k < controlled_wires.size(); k++) {
+                const std::size_t rev_wire =
+                    (num_qubits - 1) - controlled_wires[k];
+                const auto value =
+                    static_cast<std::size_t>(controlled_values[k]);
+                i = (i & ~(one << rev_wire)) | (value << rev_wire);
```

Comment:
Safe for statevector backends.

Reply:
Do you mean because `rev_wire` cannot be too large?